### DISCUSSION TITLE
How to generate an integration test binary without rust_test?

### DISCUSSION THREAD

**Discussion by tamuhey** [bazelbuild/rules_rust#913] [Q&A] [answered]
Hi, I want to know how to generate an integration test binary without `rust_test`.
My package structure is like:

```
.
├── BUILD
├── Cargo.toml
├── src
│   └── main.rs
├── tests
│   └── integration_test.rs
├── test.sh
└── WORKSPACE
```

I want to run an integration test with `test.sh` whose code is below:

```sh
setup() # setup for integration test
run_integration_test
```

Hence, I want to run `setup` before running the integration test.

I know I can get the binary of `integration_test.rs` as follows:

```
# BUILD.bazel
rust_test(
  name = "tester",
  ...
)
sh_test(
  name = "test",
  srcs = ["test.sh"],
  data = [":tester"]
)
```

However, running `bazel test ...` runs `sh_test` as well as `rust_test`, which is not desired.
I think I can solve this problem if I could get the binary of `integration_test.rs`.

How to do that?

**Comment by UebelAndre** [comment, marked as answer]
Would you be able to build a `rust_binary` with `rustc_flags = ["--test"]`? I think this would enable the libtest harness and produce a binary that would match the binary created by `rust_test` but allow you to pass it to your `sh_binary`.

**Reply by tamuhey** [reply]
It worked!! Thanks!

**Comment by hlopko** [comment]
This sounds like a weird use case, can you tell us more about why you need to do this? Are you using rust_test only so you can use `#[test]` attribute to annotate multiple tests and leave Rust test harness to do the test execution and reporting? 

Would something like this work?

```rust
#[test]
fn test() {
    run_test(|| {
        let ret_value = function_under_test();
        assert!(ret_value);
    })
}
```

On the side, Rust test harness can run tests in parallel, you can get nondeterminism if one of the tests messes up with the environment, or if one of the tests is using resources in the environment.

**Reply by tamuhey** [reply]
I want to use some tools for setting up the integration test. Of course, I can be done only with Rust code, but I want to use Bash script for simplicity.
